src/utilcode/pedecoder.cpp

```diff
@@ -1790,7 +1790,96 @@ void PEDecoder::LayoutILOnly(void *base, BOOL allowFullPE) const
 
 #endif // #ifndef DACCESS_COMPILE
 
-#ifndef FEATURE_PAL
+DWORD PEDecoder::ReadResourceDictionary(DWORD rvaOfResourceSection, DWORD rva, LPCWSTR name, BOOL *pIsDictionary) const
+{
+    *pIsDictionary = FALSE;
+
+    if (!CheckRva(rva, sizeof(IMAGE_RESOURCE_DIRECTORY)))
+    {
+        return 0;
+    }
+
+    IMAGE_RESOURCE_DIRECTORY *pResourceDirectory = (IMAGE_RESOURCE_DIRECTORY *)GetRvaData(rva);
+
+    if (pResourceDirectory->MajorVersion != 4)
+        return 0;
+
+    if (pResourceDirectory->MinorVersion != 0)
+        return 0;
+
+    // Check to see if entire resource dictionary is accessible
+    if (!CheckRva(rva + sizeof(IMAGE_RESOURCE_DIRECTORY), 
+                       (sizeof(IMAGE_RESOURCE_DIRECTORY_ENTRY) * pResourceDirectory->NumberOfNamedEntries) +
+                       (sizeof(IMAGE_RESOURCE_DIRECTORY_ENTRY) * pResourceDirectory->NumberOfIdEntries)))
+    {
+        return 0;
+    }
+
+    IMAGE_RESOURCE_DIRECTORY_ENTRY* pDirectoryEntries = (IMAGE_RESOURCE_DIRECTORY_ENTRY *)GetRvaData(rva + sizeof(IMAGE_RESOURCE_DIRECTORY));
+
+    // A fast implementation of resource lookup uses a binary search, but our needs are simple, and a linear search
+    // is easier to prove correct, so do that instead.
+    DWORD iEntryCount = (DWORD)pResourceDirectory->NumberOfNamedEntries + (DWORD)pResourceDirectory->NumberOfIdEntries;
+
+    for (DWORD iEntry = 0; iEntry < iEntryCount; iEntry++)
```

Comment:
Nit: we could limit the search to just the part that is interesting (if we're not looking for a named entry, we can start at NumberOfNamedEntries).

Reply:
I agree, but this code is simpler, and fast enough for our needs. (For instance, I believe named entries are actually first)

Reply:
You could also just take whatever is first.
You look at my rtlfilever.c in OS repo for some additional ideas.
For example, I have seen images with only a two level resource tree instead of, it is supposed to be, 3.